File: Packages/HamStationKit/Sources/Satellite/TLEParser.swift
// TLEParser.swift
// HamStationKit — Two-Line Element set parser for satellite orbit data.

import Foundation

// MARK: - TLE

/// A Two-Line Element set describing a satellite's orbital elements.
///
/// TLE format (from Celestrak):
/// ```
/// Line 0: Satellite Name
/// Line 1: 1 NNNNNC NNNNNAAA NNNNN.NNNNNNNN +.NNNNNNNN +NNNNN-N +NNNNN-N N NNNNN
/// Line 2: 2 NNNNN NNN.NNNN NNN.NNNN NNNNNNN NNN.NNNN NNN.NNNN NN.NNNNNNNNNNNNNN
/// ```
public struct TLE: Sendable, Identifiable, Equatable {
    /// NORAD catalog number.
    public let id: Int
    /// Satellite name from line 0.
    public var name: String
    /// Raw TLE line 1.
    public var line1: String
    /// Raw TLE line 2.
    public var line2: String

    // Parsed orbital elements from line 2:

    /// Orbital inclination in degrees.
    public var inclination: Double
    /// Right ascension of ascending node in degrees.
    public var raan: Double
    /// Orbital eccentricity (dimensionless, 0-1).
    public var eccentricity: Double
    /// Argument of perigee in degrees.
    public var argOfPerigee: Double
    /// Mean anomaly in degrees.
    public var meanAnomaly: Double
    /// Mean motion in revolutions per day.
    public var meanMotion: Double
    /// Epoch year (2-digit in TLE, stored as 4-digit).
    public var epochYear: Int
    /// Epoch day of year (fractional).
    public var epochDay: Double
    /// BSTAR drag term (1/earth radii).
    public var bstar: Double
    /// Revolution number at epoch.
    public var revolutionNumber: Int

    /// The epoch as a `Date`.
    public var epoch: Date {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        var components = DateComponents()
        components.year = epochYear
        components.day = 1
        components.hour = 0
        components.minute = 0
        components.second = 0
        guard let jan1 = calendar.date(from: components) else { return Date.distantPast }
        // epochDay is 1-based fractional day of year
        return jan1.addingTimeInterval((epochDay - 1.0) * 86400.0)
    }
}

// MARK: - TLEParser

/// Parses Two-Line Element set text into `TLE` values.
public struct TLEParser: Sendable {

    /// Parse a multi-satellite TLE file (Celestrak 3-line format).
    ///
    /// Each satellite occupies 3 lines: name, line 1, line 2.
    /// Blank lines and lines starting with `#` are skipped.
    public static func parse(text: String) -> [TLE] {
        let rawLines = text.components(separatedBy: .newlines)
        // Filter out blank lines and comments
        let lines = rawLines
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty && !$0.hasPrefix("#") }

        var results: [TLE] = []
        var index = 0

        while index + 2 < lines.count {
            let candidate0 = lines[index]
            let candidate1 = lines[index + 1]
            let candidate2 = lines[index + 2]

            // Check if candidate1 starts with "1 " and candidate2 starts with "2 "
            if candidate1.hasPrefix("1 ") && candidate2.hasPrefix("2 ") {
                if let tle = parseSingle(name: candidate0, line1: candidate1, line2: candidate2) {
                    results.append(tle)
                }
                index += 3
            } else {
                // Skip malformed entries
                index += 1
            }
        }

        return results
    }

    /// Parse a single TLE from its three components.
    ///
    /// - Parameters:
    ///   - name: Satellite name (line 0).
    ///   - line1: TLE line 1 (starts with "1 ").
    ///   - line2: TLE line 2 (starts with "2 ").
    /// - Returns: A parsed `TLE`, or `nil` if the input is malformed.
    public static func parseSingle(name: String, line1: String, line2: String) -> TLE? {
        // Validate line lengths (standard TLE is 69 chars per line)
        guard line1.count >= 69, line2.count >= 69 else { return nil }
        guard line1.hasPrefix("1 "), line2.hasPrefix("2 ") else { return nil }

        let l1 = Array(line1)
        let l2 = Array(line2)

        // Line 1: NORAD catalog number (columns 3-7)
        guard let catalogNumber = Int(String(l1[2...6]).trimmingCharacters(in: .whitespaces)) else { return nil }

        // Epoch year (columns 19-20) and epoch day (columns 21-32)
        guard let epochYr = Int(String(l1[18...19]).trimmingCharacters(in: .whitespaces)) else { return nil }
        guard let epochDy = Double(String(l1[20...31]).trimmingCharacters(in: .whitespaces)) else { return nil }

        // BSTAR drag term (columns 54-61): format +NNNNN-N means 0.NNNNN * 10^-N
        let bstarStr = String(l1[53...60]).trimmingCharacters(in: .whitespaces)
        let bstar = parseExponentialNotation(bstarStr)

        // Convert 2-digit year to 4-digit
        let fullYear = epochYr < 57 ? 2000 + epochYr : 1900 + epochYr

        // Line 2 parsing
        // Inclination (columns 9-16)
        guard let inclination = Double(String(l2[8...15]).trimmingCharacters(in: .whitespaces)) else { return nil }
        // RAAN (columns 18-25)
        guard let raan = Double(String(l2[17...24]).trimmingCharacters(in: .whitespaces)) else { return nil }
        // Eccentricity (columns 27-33): implied leading decimal point
        guard let eccRaw = Double("0." + String(l2[26...32]).trimmingCharacters(in: .whitespaces)) else { return nil }
        // Argument of perigee (columns 35-42)
        guard let argPerigee = Double(String(l2[34...41]).trimmingCharacters(in: .whitespaces)) else { return nil }
        // Mean anomaly (columns 44-51)
        guard let meanAnomaly = Double(String(l2[43...50]).trimmingCharacters(in: .whitespaces)) else { return nil }
        // Mean motion (columns 53-63)
        guard let meanMotion = Double(String(l2[52...62]).trimmingCharacters(in: .whitespaces)) else { return nil }
        // Revolution number at epoch (columns 64-68)
        let revNumber = Int(String(l2[63...67]).trimmingCharacters(in: .whitespaces)) ?? 0

        return TLE(
            id: catalogNumber,
            name: name.trimmingCharacters(in: .whitespacesAndNewlines),
            line1: line1,
            line2: line2,
            inclination: inclination,
            raan: raan,
            eccentricity: eccRaw,
            argOfPerigee: argPerigee,
            meanAnomaly: meanAnomaly,
            meanMotion: meanMotion,
            epochYear: fullYear,
            epochDay: epochDy,
            bstar: bstar,
            revolutionNumber: revNumber
        )
    }

    // MARK: - Private Helpers

    /// Parse TLE exponential notation like "+12345-6" or "-12345+6" into a Double.
    ///
    /// Format: `±NNNNN±E` where the value is `±0.NNNNN * 10^(±E)`.
    static func parseExponentialNotation(_ str: String) -> Double {
        var s = str.trimmingCharacters(in: .whitespaces)
        guard !s.isEmpty else { return 0.0 }

        // Handle leading sign
        var sign: Double = 1.0
        if s.first == "-" {
            sign = -1.0
            s = String(s.dropFirst())
        } else if s.first == "+" || s.first == " " {
            s = String(s.dropFirst())
        }

        // Find the exponent sign (last +/- that's not the first character)
        var mantissaStr = s
        var exponent: Double = 0.0

        // Look for +/- in the remaining string (exponent delimiter)
        if let lastSign = s.lastIndex(where: { $0 == "+" || $0 == "-" }), lastSign > s.startIndex {
            mantissaStr = String(s[s.startIndex..<lastSign])
            let expStr = String(s[lastSign...])
            exponent = Double(expStr) ?? 0.0
        }

        guard let mantissa = Double("0." + mantissaStr) else { return 0.0 }
        return sign * mantissa * pow(10.0, exponent)
    }
}

File: Packages/shared/src/satellite/tle-parser.ts
/**
 * TLE (Two-Line Element) parser for satellite orbital data.
 * Port of HamStationKit/Sources/Satellite/TLEParser.swift
 */

export interface TLE {
  id: number;
  name: string;
  inclination: number;    // degrees
  raan: number;           // Right Ascension of Ascending Node (degrees)
  eccentricity: number;   // 0-1
  argOfPerigee: number;   // degrees
  meanAnomaly: number;    // degrees
  meanMotion: number;     // revolutions per day
  epochYear: number;      // 4-digit year
  epochDay: number;       // fractional day of year
  bstar: number;          // drag term
  epoch: Date;            // computed from epochYear + epochDay
}

/**
 * Parse TLE text (3-line format: name, line1, line2).
 */
export function parseTLE(text: string): TLE[] {
  const lines = text.split('\n').map(l => l.trim()).filter(l => l.length > 0);
  const results: TLE[] = [];

  let i = 0;
  while (i < lines.length) {
    // Find a line starting with "1 " (line 1 of TLE)
    if (lines[i].startsWith('1 ') && i + 1 < lines.length && lines[i + 1].startsWith('2 ')) {
      // Line before this is the name (if it doesn't start with 1 or 2)
      const name = (i > 0 && !lines[i - 1].startsWith('1 ') && !lines[i - 1].startsWith('2 '))
        ? lines[i - 1] : 'Unknown';

      const tle = parseTLELines(name, lines[i], lines[i + 1]);
      if (tle) results.push(tle);
      i += 2;
    } else {
      i++;
    }
  }

  return results;
}

function parseTLELines(name: string, line1: string, line2: string): TLE | null {
  if (line1.length < 69 || line2.length < 69) return null;

  try {
    const catalogNumber = parseInt(line1.substring(2, 7).trim());
    const epochYr = parseInt(line1.substring(18, 20).trim());
    const epochDy = parseFloat(line1.substring(20, 32).trim());
    const bstarStr = line1.substring(53, 61).trim();

    const inclination = parseFloat(line2.substring(8, 16).trim());
    const raan = parseFloat(line2.substring(17, 25).trim());
    const eccStr = '0.' + line2.substring(26, 33).trim();
    const eccentricity = parseFloat(eccStr);
    const argPerigee = parseFloat(line2.substring(34, 42).trim());
    const meanAnomaly = parseFloat(line2.substring(43, 51).trim());
    const meanMotion = parseFloat(line2.substring(52, 63).trim());

    const fullYear = epochYr < 57 ? 2000 + epochYr : 1900 + epochYr;
    const bstar = parseExponential(bstarStr);

    // Compute epoch Date
    const jan1 = new Date(Date.UTC(fullYear, 0, 1));
    const epoch = new Date(jan1.getTime() + (epochDy - 1) * 86400000);

    return {
      id: catalogNumber,
      name: name.trim(),
      inclination, raan, eccentricity, argOfPerigee, meanAnomaly, meanMotion,
      epochYear: fullYear, epochDay: epochDy, bstar, epoch,
    };
  } catch {
    return null;
  }
}

/** Parse BSTAR-format exponential notation: "+12345-6" → 0.12345e-6 */
function parseExponential(s: string): number {
  if (!s || s.trim().length === 0) return 0;
  let cleaned = s.trim().replace(/\s+/g, '');

  // Handle format like "+12345-6" → "0.12345e-6"
  const match = cleaned.match(/^([+-]?)(\d+)([+-]\d+)$/);
  if (match) {
    const sign = match[1] === '-' ? -1 : 1;
    const mantissa = parseFloat('0.' + match[2]);
    const exponent = parseInt(match[3]);
    return sign * mantissa * Math.pow(10, exponent);
  }

  // Try direct parse
  return parseFloat(cleaned) || 0;
}

/** Well-known satellite TLEs for demo. */
export const DEMO_TLES = `ISS (ZARYA)
1 25544U 98067A   24079.50000000  .00016717  00000-0  10270-3 0  9003
2 25544  51.6400 247.4627 0006703  30.5360 325.0288 15.49815700  1234
SO-50
1 27607U 02058C   24079.00000000  .00000158  00000-0  84210-4 0  9001
2 27607  64.5567 135.2478 0025000 213.5000 146.5000 14.75500000  1234
AO-91
1 43017U 17073E   24079.00000000  .00000820  00000-0  43510-4 0  9001
2 43017  97.5000 115.0000 0013000  45.0000 315.0000 15.22000000  1234`;
